presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -385,29 +403,54 @@ private PrestoSparkQueryExecution(
                         0,
                         0,
                         serializedTaskDescriptor,
-                        new PrestoSparkTaskInputs(ImmutableMap.of(child.getFragment().getId().toString(), sparkDriverInput.iterator())),
+                        new PrestoSparkTaskInputs(ImmutableMap.of(child.getFragment().getId().toString(), sparkDriverInput.iterator()), ImmutableMap.of()),
                         taskStatsCollector));
             }
 
-            return createRdd(root).collect();
+            List<Tuple2<Integer, PrestoSparkRow>> result = createRdd(root, broadcastsToDestroy).collect();
+            broadcastsToDestroy.forEach(Broadcast::destroy);
+            return result;
         }
 
-        private JavaPairRDD<Integer, PrestoSparkRow> createRdd(SubPlan subPlan)
+        private JavaPairRDD<Integer, PrestoSparkRow> createRdd(SubPlan subPlan, List<Broadcast<?>> broadcasts)
         {
-            PlanFragment fragment = subPlan.getFragment();
-            Map<PlanFragmentId, JavaPairRDD<Integer, PrestoSparkRow>> rddInputs = subPlan.getChildren().stream()
-                    .collect(toImmutableMap(children -> children.getFragment().getId(), this::createRdd));
+            ImmutableMap.Builder<PlanFragmentId, JavaPairRDD<Integer, PrestoSparkRow>> rddInputs = ImmutableMap.builder();
+            ImmutableMap.Builder<PlanFragmentId, Broadcast<List<PrestoSparkSerializedPage>>> broadcastInputs = ImmutableMap.builder();
+            for (SubPlan child : subPlan.getChildren()) {
+                PlanFragment childFragment = child.getFragment();
+                if (childFragment.getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_BROADCAST_DISTRIBUTION)) {
+                    List<Broadcast<?>> broadcastsToDestroy = new ArrayList<>();
+                    JavaPairRDD<Integer, PrestoSparkRow> childRdd = createRdd(child, broadcastsToDestroy);
+                    List<Tuple2<Integer, PrestoSparkRow>> broadcastRows = childRdd.collect();
+                    // TODO: Transform rows to pages on executors (using `RDD#map` function)
+                    // TODO: Transforming it on coordinator results in 2x memory utilization as both,
+                    // TODO: rows and pages have to be kept in memory all at the same time
+                    Iterator<Page> pagesIterator = transformRowsToPages(transform(broadcastRows.iterator(), Tuple2::_2), childFragment.getTypes());
+                    Iterator<PrestoSparkSerializedPage> serializedPagesIterator = transform(transform(pagesIterator, pagesSerde::serialize), PrestoSparkUtils::toPrestoSparkSerializedPage);
+                    List<PrestoSparkSerializedPage> serializedPages = new ArrayList<>();
+                    serializedPagesIterator.forEachRemaining(serializedPages::add);
+                    Broadcast<List<PrestoSparkSerializedPage>> broadcast = sparkContext.broadcast(serializedPages);
```

Comment:
Similar to https://github.com/apache/spark/blob/5d5866be12259c40972f7404f64d830cab87401f/sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala#L109-L112 , it might help to have some sanity check on data being broadcasted to prevent Driver from OOM. This can be left as a future work.

Reply:
I don't like checking the row size, and I don't think there's a way to figure out the actual size in bytes in advance (@sameeragarwal ?). I don't like limiting the number of rows, as the row size could be anywhere from several bytes to hundred of kilobytes. Thus it doesn't give us much.

Reply:
> I don't like checking the row size

Do you mean the size of byte array after serializing page?  Doesn't the array length of `PrestoSparkSerializedPage#bytes` work?

Reply:
It is too late at that step. The data is already on the driver. If we expect executors to have less memory available than a driver we can add a check. But that might be cluster specific.

Reply:
@arhimondr : I see. So it won't protect us from OOM. It may still help with organic growth of the broadcast table (e.g. the broadcast table gradually grows from 10G to 20G...). So the value is not that large.